src/com/ms/silverking/cloud/dht/client/impl/AsyncOperationImpl.java

```diff
@@ -27,446 +28,490 @@
-     * @param listener    completion listener
-     */
-    public void addListener(AsyncOperationListener listener) {
-        addListener(listener, OperationState.SUCCEEDED, OperationState.FAILED);
+
+    @Override
+    public void doWork(Pair<AsyncOperationImpl, Set<AsyncOperationListener>> p) {
+      p.getV1().notifyListeners(p.getV2());
     }
-    
-    /**
-     * Adds multiple completion listeners. For any listener that is already complete, the callback will be 
-     * immediately executed, possibly in the calling thread.
-     * Equivalent to addListeners(listeners, OperationState.SUCCEEDED, OperationState.FAILED)
-     * @param listeners    update listeners
-     */
-    public void addListeners(Iterable<AsyncOperationListener> listeners) {
-        addListeners(listeners, OperationState.SUCCEEDED, OperationState.FAILED);
-    }    
-    
-    /**
-     * Adds an operation listener. If the operation is already complete, the callback will be 
-     * immediately executed, possibly in the calling thread.
-     * @param listener    update listener
-     * @param listenStates    states to generate updates for
-     */
-    public void addListener(AsyncOperationListener listener, OperationState... listenStates) {
-        EnumSet<OperationState>    _listenStates;
-        OperationState            opState;
-
-        _listenStates = CollectionUtil.arrayToEnumSet(listenStates);
-        opState = getState();
-        lock.lock();
+  }
+
+  public boolean waitForCompletion(long timeout, TimeUnit unit) throws OperationException {
+    long relativeDeadlineMillis;
+    long absoluteDeadlineMillis;
+
+    relativeDeadlineMillis = TimeUnit.MILLISECONDS.convert(timeout, unit);
+    absoluteDeadlineMillis = SystemTimeUtil.skSystemTimeSource.absTimeMillis() + relativeDeadlineMillis;
+    lock.lock();
+    try {
+      while (!result.isComplete()) {
         try {
-            // Trigger immediate updates for completion, but only for completion
-            if (opState != OperationState.INCOMPLETE && _listenStates.contains(opState)) {
-                listener.asyncOperationUpdated(this);
+          long millisToDeadline;
+
+          millisToDeadline = absoluteDeadlineMillis - SystemTimeUtil.skSystemTimeSource.absTimeMillis();
+          if (millisToDeadline > 0) {
+            Log.fine("activeOp awaiting ", this);
+            if (!spin) {
+              cv.await(millisToDeadline, TimeUnit.MILLISECONDS);
             } else {
-                if (listeners == null) {
-                    listeners = new HashSet<>();
-                }
-                listeners.add(new Pair<>(listener, _listenStates));
-                opsWithListeners.put(getUUID(), this);
+              cv.awaitNanos(spinDurationNanos);
             }
-        } finally {
-            lock.unlock();
+          } else {
+            debugTimeout();
+            return false;
+          }
+        } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)